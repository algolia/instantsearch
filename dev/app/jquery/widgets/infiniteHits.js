/* eslint-disable import/default */
import instantsearch from '../../../../index';

const renderFn = ({
  hits,
  showMore,
  isLastPage,
  widgetParams: { containerNode },
}) => {
  const hitsHTML = hits.map(
    hit => `
    <div class="hit">
      <div class="media">
        <a
          href="${hit.url}"
          class="pull-left"
        >
          <img
            class="media-object"
            src="${hit.image}"
          />
        </a>

        <div class="media-body">
          <h3 class="pull-right text-right text-info">$${hit.price}</h3>
          <h4>${hit._highlightResult.name.value}</h4>
          <p>${hit._highlightResult.description.value}</p>

          ${
            hit.free_shipping
              ? '<span class="badge pull-right">Free shipping</span>'
              : ''
          }
        </div>
      </div>
    </div>
  `
  );

  containerNode.find('button[data-show-more]').off('click');
  containerNode.html(hitsHTML);

  if (!isLastPage) {
    containerNode.append(`
      <button class="btn btn-default" data-show-more>
        Show more
      </button>
    `);

    containerNode.find('button[data-show-more]').on('click', e => {
      e.preventDefault();
      showMore();
    });
  }
};

export default instantsearch.connectors.connectInfiniteHits(renderFn);
