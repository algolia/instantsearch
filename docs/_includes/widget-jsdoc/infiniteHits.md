<h4 class="no-toc">Parameters</h4>
<p class="attr-name">
<span class='attr-required'>`options.container`<span class="show-description">…</span></span>
  <span class="attr-infos">(<code>string</code> &#124; <code>DOMElement</code>)</span>
</p>
<p class="attr-description important">CSS Selector or DOMElement to insert the widget</p>
<p class="attr-name">
<span class='attr-optional important'>`options.hitsPerPage`<span class="show-description">…</span></span>
  <span class="attr-infos">Default:<code class="attr-default">20</code>(<code>number</code>)</span>
</p>
<p class="attr-description important">The number of hits to display per page</p>
<p class="attr-name">
<span class='attr-optional'>`options.templates`<span class="show-description">…</span></span>
  <span class="attr-infos">(<code>Object</code>)</span>
</p>
<p class="attr-description">Templates to use for the widget</p>
<p class="attr-name">
<span class='attr-optional'>`options.templates.empty`<span class="show-description">…</span></span>
  <span class="attr-infos">Default:<code class="attr-default">&quot;&quot;</code>(<code>string</code> &#124; <code>function</code>)</span>
</p>
<p class="attr-description">Template to use when there are no results.</p>
<p class="attr-name">
<span class='attr-optional'>`options.templates.item`<span class="show-description">…</span></span>
  <span class="attr-infos">Default:<code class="attr-default">&quot;&quot;</code>(<code>string</code> &#124; <code>function</code>)</span>
</p>
<p class="attr-description">Template to use for each result. This template will receive an object containing a single record.</p>
<p class="attr-name">
<span class='attr-optional'>`options.showMoreLabel`<span class="show-description">…</span></span>
  <span class="attr-infos">Default:<code class="attr-default">&quot;Show more results&quot;</code>(<code>string</code>)</span>
</p>
<p class="attr-description">label used on the show more button</p>
<p class="attr-name">
<span class='attr-optional'>`options.transformData`<span class="show-description">…</span></span>
  <span class="attr-infos">(<code>Object</code>)</span>
</p>
<p class="attr-description">Method to change the object passed to the templates</p>
<p class="attr-name">
<span class='attr-optional'>`options.transformData.empty`<span class="show-description">…</span></span>
  <span class="attr-infos">(<code>function</code>)</span>
</p>
<p class="attr-description">Method used to change the object passed to the `empty` template</p>
<p class="attr-name">
<span class='attr-optional'>`options.transformData.item`<span class="show-description">…</span></span>
  <span class="attr-infos">(<code>function</code>)</span>
</p>
<p class="attr-description">Method used to change the object passed to the `item` template</p>
<p class="attr-name">
<span class='attr-optional'>`options.cssClasses`<span class="show-description">…</span></span>
  <span class="attr-infos">(<code>Object</code>)</span>
</p>
<p class="attr-description">CSS classes to add</p>
<p class="attr-name">
<span class='attr-optional'>`options.cssClasses.root`<span class="show-description">…</span></span>
  <span class="attr-infos">(<code>string</code> &#124; <code>Array.&lt;string&gt;</code>)</span>
</p>
<p class="attr-description">CSS class to add to the wrapping element</p>
<p class="attr-name">
<span class='attr-optional'>`options.cssClasses.empty`<span class="show-description">…</span></span>
  <span class="attr-infos">(<code>string</code> &#124; <code>Array.&lt;string&gt;</code>)</span>
</p>
<p class="attr-description">CSS class to add to the wrapping element when no results</p>
<p class="attr-name">
<span class='attr-optional'>`options.cssClasses.item`<span class="show-description">…</span></span>
  <span class="attr-infos">(<code>string</code> &#124; <code>Array.&lt;string&gt;</code>)</span>
</p>
<p class="attr-description">CSS class to add to each result</p>

<p class="attr-legend">* <span>Required</span></p>
