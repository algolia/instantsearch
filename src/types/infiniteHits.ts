import { SearchParameters } from 'algoliasearch-helper';
import { Hits } from './';

export type InfiniteHitsCachedHits = {
  [page: number]: Hits;
};

type Read = ({
  state,
}: {
  state: Partial<SearchParameters>;
}) => InfiniteHitsCachedHits | null;

type Write = ({
  state,
  hits,
}: {
  state: Partial<SearchParameters>;
  hits: InfiniteHitsCachedHits;
}) => void;

export type InfiniteHitsCache = {
  read: Read;
  write: Write;
};
