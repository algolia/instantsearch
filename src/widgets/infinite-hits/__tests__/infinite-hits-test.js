import algoliasearchHelper from 'algoliasearch-helper';
import infiniteHits from '../infinite-hits';

describe('infiniteHits call', () => {
  it('throws an exception when no container', () => {
    expect(infiniteHits).toThrow();
  });
});

describe('infiniteHits()', () => {
  let ReactDOM;
  let container;
  let widget;
  let results;
  let helper;

  beforeEach(() => {
    helper = algoliasearchHelper({});
    helper.search = jest.fn();

    ReactDOM = { render: jest.fn() };
    infiniteHits.__Rewire__('render', ReactDOM.render);

    container = document.createElement('div');
    widget = infiniteHits({
      container,
      escapeHits: true,
      cssClasses: { root: ['root', 'cx'] },
    });
    widget.init({ helper, instantSearchInstance: {} });
    results = { hits: [{ first: 'hit', second: 'hit' }] };
  });

  it('It does have a specific configuration', () => {
    expect(widget.getConfiguration()).toEqual({
      highlightPostTag: '__/ais-highlight__',
      highlightPreTag: '__ais-highlight__',
    });
  });

  it('calls twice ReactDOM.render(<Hits props />, container)', () => {
    const state = { page: 0 };
    widget.render({ results, state });
    widget.render({ results, state });

    expect(ReactDOM.render).toHaveBeenCalledTimes(
      2,
      'ReactDOM.render called twice'
    );
    expect(ReactDOM.render.mock.calls[0][0]).toMatchSnapshot();
    expect(ReactDOM.render.mock.calls[0][1]).toEqual(container);
    expect(ReactDOM.render.mock.calls[1][0]).toMatchSnapshot();
    expect(ReactDOM.render.mock.calls[1][1]).toEqual(container);
  });

  it('renders transformed items', () => {
    const state = { page: 0 };
    widget = infiniteHits({
      container,
      transformItems: items =>
        items.map(item => ({ ...item, transformed: true })),
    });

    widget.init({ helper, instantSearchInstance: {} });
    widget.render({
      results,
      state,
      instantSearchInstance: {},
    });

    expect(ReactDOM.render.mock.calls[0][0]).toMatchSnapshot();
  });

  it('if it is the last page, then the props should contain isLastPage true', () => {
    const state = { page: 0 };
    widget.render({
      results: { ...results, page: 0, nbPages: 2 },
      state,
    });
    widget.render({
      results: { ...results, page: 1, nbPages: 2 },
      state,
    });

    expect(ReactDOM.render).toHaveBeenCalledTimes(
      2,
      'ReactDOM.render called twice'
    );
    expect(ReactDOM.render.mock.calls[0][0]).toMatchSnapshot();
    expect(ReactDOM.render.mock.calls[0][1]).toEqual(container);
    expect(ReactDOM.render.mock.calls[1][0]).toMatchSnapshot();
    expect(ReactDOM.render.mock.calls[1][1]).toEqual(container);
  });

  it('does not accept allItems templates', () => {
    expect(() =>
      infiniteHits({ container, templates: { allItems: '' } })
    ).toThrow();

    expect(() =>
      infiniteHits({ container, templates: { allItems: 'not empty' } })
    ).toThrow();

    expect(() =>
      infiniteHits({ container, templates: { allItems: () => 'template' } })
    ).toThrow();
  });

  it('updates the search state properly when showMore is called', () => {
    expect(helper.state.page).toBe(0);

    widget.showMore();

    expect(helper.state.page).toBe(1);
    expect(helper.search).toHaveBeenCalledTimes(1);
  });

  afterEach(() => {
    infiniteHits.__ResetDependency__('render');
    infiniteHits.__ResetDependency__('defaultTemplates');
  });
});
