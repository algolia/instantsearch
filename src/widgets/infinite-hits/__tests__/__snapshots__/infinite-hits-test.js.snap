// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`infiniteHits() calls twice ReactDOM.render(<Hits props />, container) 1`] = `
<InfiniteHits
  cssClasses={
    Object {
      "empty": "ais-infinite-hits__empty",
      "item": "ais-infinite-hits--item",
      "root": "ais-infinite-hits root cx",
      "showmore": "ais-infinite-hits--showmore",
      "showmoreButton": "ais-infinite-hits--showmoreButton",
    }
  }
  hits={
    Array [
      Object {
        "first": "hit",
        "second": "hit",
      },
    ]
  }
  isLastPage={false}
  results={
    Object {
      "hits": Array [
        Object {
          "first": "hit",
          "second": "hit",
        },
      ],
    }
  }
  showMore={[Function]}
  showMoreLabel="Show more results"
  templateProps={
    Object {
      "templates": Object {
        "empty": "No results",
        "item": [Function],
      },
      "templatesConfig": undefined,
      "transformData": undefined,
      "useCustomCompileOptions": Object {
        "empty": false,
        "item": false,
      },
    }
  }
/>
`;

exports[`infiniteHits() calls twice ReactDOM.render(<Hits props />, container) 2`] = `
<InfiniteHits
  cssClasses={
    Object {
      "empty": "ais-infinite-hits__empty",
      "item": "ais-infinite-hits--item",
      "root": "ais-infinite-hits root cx",
      "showmore": "ais-infinite-hits--showmore",
      "showmoreButton": "ais-infinite-hits--showmoreButton",
    }
  }
  hits={
    Array [
      Object {
        "first": "hit",
        "second": "hit",
      },
    ]
  }
  isLastPage={false}
  results={
    Object {
      "hits": Array [
        Object {
          "first": "hit",
          "second": "hit",
        },
      ],
    }
  }
  showMore={[Function]}
  showMoreLabel="Show more results"
  templateProps={
    Object {
      "templates": Object {
        "empty": "No results",
        "item": [Function],
      },
      "templatesConfig": undefined,
      "transformData": undefined,
      "useCustomCompileOptions": Object {
        "empty": false,
        "item": false,
      },
    }
  }
/>
`;

exports[`infiniteHits() if it is the last page, then the props should contain isLastPage true 1`] = `
<InfiniteHits
  cssClasses={
    Object {
      "empty": "ais-infinite-hits__empty",
      "item": "ais-infinite-hits--item",
      "root": "ais-infinite-hits root cx",
      "showmore": "ais-infinite-hits--showmore",
      "showmoreButton": "ais-infinite-hits--showmoreButton",
    }
  }
  hits={
    Array [
      Object {
        "first": "hit",
        "second": "hit",
      },
    ]
  }
  isLastPage={false}
  results={
    Object {
      "hits": Array [
        Object {
          "first": "hit",
          "second": "hit",
        },
      ],
      "nbPages": 2,
      "page": 0,
    }
  }
  showMore={[Function]}
  showMoreLabel="Show more results"
  templateProps={
    Object {
      "templates": Object {
        "empty": "No results",
        "item": [Function],
      },
      "templatesConfig": undefined,
      "transformData": undefined,
      "useCustomCompileOptions": Object {
        "empty": false,
        "item": false,
      },
    }
  }
/>
`;

exports[`infiniteHits() if it is the last page, then the props should contain isLastPage true 2`] = `
<InfiniteHits
  cssClasses={
    Object {
      "empty": "ais-infinite-hits__empty",
      "item": "ais-infinite-hits--item",
      "root": "ais-infinite-hits root cx",
      "showmore": "ais-infinite-hits--showmore",
      "showmoreButton": "ais-infinite-hits--showmoreButton",
    }
  }
  hits={
    Array [
      Object {
        "first": "hit",
        "second": "hit",
      },
    ]
  }
  isLastPage={true}
  results={
    Object {
      "hits": Array [
        Object {
          "first": "hit",
          "second": "hit",
        },
      ],
      "nbPages": 2,
      "page": 1,
    }
  }
  showMore={[Function]}
  showMoreLabel="Show more results"
  templateProps={
    Object {
      "templates": Object {
        "empty": "No results",
        "item": [Function],
      },
      "templatesConfig": undefined,
      "transformData": undefined,
      "useCustomCompileOptions": Object {
        "empty": false,
        "item": false,
      },
    }
  }
/>
`;

exports[`infiniteHits() renders transformed items 1`] = `
<InfiniteHits
  cssClasses={
    Object {
      "empty": "ais-infinite-hits__empty",
      "item": "ais-infinite-hits--item",
      "root": "ais-infinite-hits",
      "showmore": "ais-infinite-hits--showmore",
      "showmoreButton": "ais-infinite-hits--showmoreButton",
    }
  }
  hits={
    Array [
      Object {
        "first": "hit",
        "second": "hit",
        "transformed": true,
      },
    ]
  }
  isLastPage={false}
  results={
    Object {
      "hits": Array [
        Object {
          "first": "hit",
          "second": "hit",
          "transformed": true,
        },
      ],
    }
  }
  showMore={[Function]}
  showMoreLabel="Show more results"
  templateProps={
    Object {
      "templates": Object {
        "empty": "No results",
        "item": [Function],
      },
      "templatesConfig": undefined,
      "transformData": undefined,
      "useCustomCompileOptions": Object {
        "empty": false,
        "item": false,
      },
    }
  }
/>
`;
